### DISCUSSION TITLE
Saleor multi-vendor

### DISCUSSION THREAD

**Discussion by moshegorin** [saleor/saleor#6905] [Ideas]
We are looking to build multi-vendor functionality on Saleor.

At the basic level this would mean every product/variation could be assigned to multiple vendors.  Each vendor would have their own price and stock levels.

On front-end customers should be able to browse by category (and see cheapest vendor by default), or browse each vendor's store. 

We have our own python team but would like some help. Can anyone help us with this?

Would like some pointers on best way to do it.  

If necessary we can also pay hourly.

**Comment by karolkielecki** [comment]
Hi there! 

Saleor has not been designed to support a multi-vendor functionality. It's a big change. Basically, you would need to develop the platform from the beginning. You would need to apply many changes to the core, dashboard and storefront.

**Reply by moshegorin** [reply]
We are not looking full-blown multi-vendor like Amazon and Etsy

All we need is that each product/variation of product can be associated with a vendor which has the price, stock level and pos_id

For example the response from the product API could look something like this:

```{
  "id": 972352,
  "name": "Milk",
  "sku": "07862614574",
  "brand": "Dairy Land",
  "size": "64oz",
  "vendors": [
    {
      "pos_id": 1072,
      "vendor": "Vendor A",
      "price": 12.99,
      "stock": true
    },
    {
      "pos_id": 9051,
      "vendor": "Vendor B",
      "price": 11.99,
      "stock": false
    },
    {
      "pos_id": null,
      "vendor": "Vendor C",
      "price": 12.49,
      "stock": true
    }
  ]
}
```

In the SQL we are thinking to add a table `product_vendor` with fields for `product_id`, `product_productvariant_id`, `vendor_id`, `price`, `stock`, `pos_id`

We understand we would have to change the API, Dashboard and Storefront but I don't think the changes should be so significant..

Please let me know what you think and if anyone can help.  Would be interested in making code public if there is interest.

**Reply by korolevpavel** [reply]
I see the difficulty in separating balances by prices. Now the balances are stored in warehouses, but if you want to use several suppliers, you need to store the balances for each invoice.

**Reply by moshegorin** [reply]
We don't need use the warehouse, supplier or invoice features of saleor right now.  Please message us through https://gorinsystems.com/contact if you think you can help (can pay)

**Reply by 54331241** [reply]
@korolevpavel you anyhow need to store the prices in the invoice - since prices can change and we need to store in the invoice the price they paid.

**Comment by moshegorin** [comment]
@JZinkl suggested we use "Collections" 

E.x. we have "Product A" and Vendors 1 and 2.

"Product A" would be title of a colleciton, and inside the collection would be Products A1 and Products A2 - one for Vendor 1 one for Vendor 2.

The in the "Product A" collection meta data we would store `preferred_vendor` and `price` which would be calculated nightly - i.e. the vendor we want to show for this Product.  Similar to when searching on Amazon it shows the preferred vendor and preferred vendor's price on the main product page.

This is a great solution but the big problem is how to organize collections into categories?  I see that products can be organized into categories but not collections..

does anyone have any ideas?

**Reply by mikedizon** [reply]
i've got it working with django-tenants, however, tests are failing. i'm sure someone with more experience with pytest and django-tenants can get them to work.
